up-l2/api.adoc

```diff
@@ -0,0 +1,62 @@
+= Messaging Pattern APIs
+:toc:
+:sectnums:
+
+The key words "*MUST*", "*MUST NOT*", "*REQUIRED*", "*SHALL*", "*SHALL NOT*", "*SHOULD*", "*SHOULD NOT*", "*RECOMMENDED*", "*MAY*", and "*OPTIONAL*" in this document are to be interpreted as described in https://www.rfc-editor.org/info/bcp14[IETF BCP14 (RFC2119 & RFC8174)]
+
+----
+SPDX-FileCopyrightText: 2023 Contributors to the Eclipse Foundation
+
+See the NOTICE file(s) distributed with this work for additional
+information regarding copyright ownership.
+
+This program and the accompanying materials are made available under
+the terms of the Apache License Version 2.0 which is available at
+https://www.apache.org/licenses/LICENSE-2.0
+ 
+SPDX-FileType: DOCUMENTATION
+SPDX-License-Identifier: Apache-2.0
+----
+
+== Overview
+
+The Messaging Pattern APIs provide the common language specific interfaces for the pub/sub, notification, and RPC patterns supported by uProtocol implementations. The interfaces are declared in the link:../languages.adoc[language-specific uProtocol libraries] and as such are merely described at a high level in this document to allow the language-specific libraries to declare the interface signatures using appropriate programming language constructs.
+
+
+== Interfaces for Messaging Patterns
+
+.Messaging Pattern APIs
+[#messaging-apis, cols="1,1,3",options="header"]
+|===
+|Messaging Pattern | Interface Name | Purpose
+
+| Pub/Sub
+| `*Subscriber*`
+| APIs to subscribe and unsubscribe to topics, as well as setup handlers (`UListeners`) to receive the published messages.
+
+| Pub/Sub
+| `*Publisher*`
+| APIs to publish messages to topics.
+
+| Notification
+| `*Notifier*`
+| APIs to send notifications as well as register/unregister notification handlers (`UListeners`).
+
+| RPC
+| `*RpcServer*`
+| APIs to register handlers that will be dispatched to process incoming service requests and automatically send back the response when processing of the request is completed.
+
+| RPC
+| `*RpcClient*`
+| APIs to invoke service methods and process the outcome of the service method invocation.
+
+|===
+
+
+* link:../languages.adoc[language-specific uProtocol libraries] *MUST* provide a _default_ implementation of the <<messaging-apis>> based on the link:../up-l1/README.adoc[Transport Level API].
+
+* *MUST* follow the link:../basics/error_model.adoc[uProtocol Error Model] for error handling for the respective APIs.
+
+* link:../upclient.adoc[uPClient library] *MAY* provide additional implementations of <<messaging-apis>> which, for example, take advantage of specific features of the underlying communication protocol.
```

Comment:
I think I found my answer to [this question](https://github.com/eclipse-uprotocol/up-spec/pull/150#discussion_r1608975915) here! So if desired, depending on the particulars of how a language chooses to implement uP-L2 (as an interface or as a class), then it would be possible for e.g. the Zenoh client to have its own way of doing RpcClient that doesn't rest upon the foundations of uP-L1.

Is that correct, @stevenhartley?

Reply:
Yes that can be true so long as L1 still works for all the patterns as dispatchers (streamer) will only use L1